File: leetcode-algorithms/083. Remove Duplicates from Sorted List/Solution.cpp
// Given a sorted linked list, delete all duplicates such that each element appear only once.
//
// Example 1:
//
// Input: 1->1->2
// Output: 1->2
// Example 2:
//
// Input: 1->1->2->3->3
// Output: 1->2->3
class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        ListNode* p = new ListNode(0);
        ListNode* cp = p;
        ListNode* next;

        while (head) {
            p->next = head;
            p = head;
            next = head;
            while ((next != nullptr) && (next->val == head->val)) next = next->next;

            head = next;
        }
        p->next = nullptr;
        return cp->next;
    }
};

File: leetcode-algorithms/083. Remove Duplicates from Sorted List/Solution.java
class Solution {
    public ListNode deleteDuplicates(ListNode head) {

        ListNode dummy = new ListNode(1);
        dummy.next = head;
        ListNode cp = dummy;
        ListNode p;
        while (head != null) {
            p = head;
            while ((p != null) && (p.val == head.val)) {
                p = p.next;
            }

            dummy.next = head;
            if (p == null) {
                head.next = null;
                break;
            }
            dummy = head;
            head = p;
        }

        return cp.next;
    }
}
